### PR TITLE
feat(pt_expt): add LMDB dataset support

### PR DESCRIPTION
**Pull request by wanghan-iapcm**
## Summary

Mirror the pt backend's LMDB plumbing (#5283) for `pt_expt` so users can train against `.lmdb` directories on the new torch.export-based stack.

- The frame-collation step is implemented once in `dpmodel` via `array_api_compat` (so future jax/paddle backends inherit it).
- pt's existing `_collate_lmdb_batch` is refactored to delegate to that shared helper — behavior preserved.
- pt_expt gets a thin `LmdbDataSystem` adapter that exposes the small surface its trainer touches (`get_batch`, `add_data_requirements`, `get_nsystems`).
- pt_expt's `get_trainer` routes scalar LMDB paths via `is_lmdb()` for both single-task and multi-task data sections.

## Files

- `deepmd/dpmodel/utils/lmdb_data.py` — new `collate_lmdb_frames()`.
- `deepmd/pt/utils/lmdb_dataset.py` — `_collate_lmdb_batch` refactored to delegate.
- `deepmd/pt_expt/utils/lmdb_dataset.py` (new) — `LmdbDataSystem` adapter.
- `deepmd/pt_expt/entrypoints/main.py` — `_build_data_system()` helper with LMDB routing.
- `source/tests/pt_expt/test_lmdb_training.py` (new) — synthetic LMDB fixture, `get_batch` shape check, end-to-end training-loop test.

## Known limitations

- `mixed_batch=True` not supported (matches pt; both raise `NotImplementedError`).
- DDP wiring deferred — `DistributedSameNlocBatchSampler` exists in dpmodel; needs a multi-GPU CI path to actually exercise.
- pt_expt's stat collection treats the LMDB as one logical system (`get_nsystems()` returns 1); per-system reweighting still happens through `SameNlocBatchSampler`'s `block_targets`.

## Test plan

- [x] `pytest source/tests/pt_expt/test_lmdb_training.py` — 4 passed
- [x] `pytest source/tests/pt/test_lmdb_dataloader.py` — 48 passed (pt collate refactor regression)
- [x] `pytest source/tests/common/dpmodel/test_lmdb_data.py` — 50 passed (dpmodel module-level regression)
- [x] `pytest source/tests/consistent/test_lmdb_data.py` — 7 passed (cross-backend dpmodel ↔ pt consistency)

## Summary by CodeRabbit

* **New Features**
  * LMDB-backed datasets are supported end-to-end in training: automatic routing, seed propagation, and a trainer-facing data system for sampling and batch retrieval.
  * Centralized collation produces consistent NumPy/Tensor batches from LMDB frames, preserving metadata and handling variable-length fields.

* **Tests**
  * Added unit and end-to-end tests verifying LMDB integration, batch shapes, epoch iteration, data-requirement passthrough, and trainer compatibility.

### CODE REVIEW

File: deepmd/pt_expt/utils/lmdb_dataset.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    LmdbDataReader,
+    SameNlocBatchSampler,
+    collate_lmdb_frames,
+    compute_block_targets,
+)
+from deepmd.utils.data import (
+    DataRequirementItem,
+)
+
+log = logging.getLogger(__name__)
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/pt_expt/utils/lmdb_dataset.py]: Fixed in 5bc00614 — dropped the unused `log` global and the `import logging` line from `deepmd/pt_expt/utils/lmdb_dataset.py`.

File: deepmd/pt_expt/entrypoints/main.py [hunk 2]
```diff
@@ -35,6 +41,41 @@
...
+) -> DeepmdDataSystem | LmdbDataSystem:
+    """Build a data system from dataset config, routing LMDB paths to LmdbDataSystem.
+
+    A scalar ``systems`` value pointing at an LMDB directory triggers the
+    LMDB adapter; otherwise we fall through to the legacy
+    :class:`DeepmdDataSystem` path with system expansion.
+    """
+    systems_raw = dataset_params["systems"]
+    if isinstance(systems_raw, str) and is_lmdb(systems_raw):
+        return LmdbDataSystem(
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/pt_expt/entrypoints/main.py]: Good catch — this was a real bug, not just a theoretical one. Fixed in 5bc00614. Added a `_get_neighbor_stat_data` helper next to `_build_data_system` that mirrors the pt backend's pattern at `deepmd/pt/entrypoints/main.py:372-381`: `is_lmdb(systems)` routes to `make_neighbor_stat_data(...)`, otherwise fall through to `get_data(...)`. Applied to both single-task and multi-task branches of the CLI `train()` flow.

File: source/tests/pt_expt/test_lmdb_training.py [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+        config = normalize(config)
+
+        cwd = os.getcwd()
+        os.chdir(self.tmpdir)
+        try:
+            trainer = get_trainer(config)
+            self.assertIsInstance(trainer.training_data, LmdbDataSystem)
+            trainer.run()
+        finally:
+            os.chdir(cwd)
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/tests/pt_expt/test_lmdb_training.py:245]: Dismissing — the "60s training test timeout" guideline referenced here doesn't exist in this repo. `rg 'pytest.mark.timeout' source/tests/` returns zero hits, and there's no such rule in CLAUDE.md or any `.github/skills/` file. The test already completes in ~4s on CI and ~3.5s locally; if a future change hangs it, the runner-level timeout will catch that. Happy to revisit if the guideline actually gets added to the repo.

**Comment by wanghan-iapcm** [comment]
@coderabbitai addressing the nitpick comments from the review body:

**Fixed in 5bc00614**:

- **`deepmd/pt_expt/utils/lmdb_dataset.py:119-121`** (private `_type_map` access): added a public `type_map` property on `LmdbDataReader` and forwarded to it. Adapter no longer touches `_reader._type_map`.
- **`deepmd/pt_expt/entrypoints/main.py:55-76`** (list-form LMDB): `_build_data_system` now raises a clear `ValueError` when a list-form `systems` contains any LMDB path. Went with the explicit-error option over silent unwrapping — unwrapping a length-1 list sets a precedent for ambiguous configs.

**Dismissing**:

- **`source/tests/pt_expt/test_lmdb_training.py:238-245`** (`monkeypatch.chdir`): the test is `unittest.TestCase`-based; `monkeypatch` is a pytest fixture and would require converting the whole class. Not worth the churn to replace a 4-line try/finally.
- **`deepmd/pt_expt/utils/lmdb_dataset.py:87-100`** (`StopIteration` → `RuntimeError`): an empty LMDB fails much earlier — `LmdbDataReader.__init__` parses metadata that requires `nframes > 0`, and `SameNlocBatchSampler` would have no nloc groups. `get_batch` can't be reached with a truly empty dataset.
- **`deepmd/dpmodel/utils/lmdb_data.py:611-643`** (`find_*` invariant): the invariant holds by construction — `LmdbDataReader.__getitem__` populates `find_*` deterministically from `data_requirements` (either `1.0` always or `0.0` always per key). No caller produces mixed-availability frames. Adding a per-batch assert would fire on hot-path collation for a contract already enforced at the reader.

File: deepmd/pt_expt/entrypoints/main.py [hunk 4]
```diff
@@ -35,6 +41,69 @@
...
+    npy/HDF5 directories.
+    """
+    systems_raw = dataset_params.get("systems")
+    if isinstance(systems_raw, str) and is_lmdb(systems_raw):
+        from deepmd.dpmodel.utils.lmdb_data import (
+            make_neighbor_stat_data,
+        )
+
+        return make_neighbor_stat_data(systems_raw, type_map)
+    return get_data(dataset_params, 0, type_map, None)
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/pt_expt/entrypoints/main.py:83]: Good catch — valid concern, fixed in afa4054c. Extracted a `_detect_lmdb_path(systems_raw)` helper that (a) returns the path for scalar LMDB strings, (b) raises the same `ValueError` for list-form `systems` containing any LMDB path, and (c) returns `None` otherwise. Both `_get_neighbor_stat_data` and `_build_data_system` now call it, so the neighbor-stat entry point and the trainer entry point fail fast with identical messages.

**Review by iProzd — APPROVED** [review]
